lixPackageSets.{lix_2_93,git}: init

alois31 opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

alois31 committed:
lixPackageSets: support splicing

alois31 committed:
lixPackageSets.git: init at 2.94.0-pre-20250509_dcb0a97000d5

alois31 committed:
lixPackageSets.lix_2_93: init at 2.93.0

alois31 commented:
The checks performed are as follows:
* built on x86_64-linux up to: `lixPackageSets.git.lix.passthru.tests`, `lixPackageSets.git.nix-eval-jobs`, `pkgsStatic.lixPackageSets.git.lix` `pkgsCross.aarch64-multiplatform.lixPackageSets.git.nix-eval-jobs`
* built on aarch64-linux up to: `lixPackageSets.git.nix-eval-jobs`, `pkgsStatic.lixPackageSets.git.lix`
* evaluated my entire configuration with it

Updates roughly once a week with the same testing are planned. Please tell me if I missed some things I should test and I can add it to the list. (Unfortunately anything Darwin is not possible for me since I do not have appropriate hardware.)

lf- reviewed (commented):
overall lgtm, I've pointed to a few tiny nits but we can get this going pretty soon!

alois31 commented on pkgs/tools/package-management/lix/common-lix.nix:
```diff
@@ -161,6 +169,8 @@ stdenv.mkDerivation (finalAttrs: {
       doxygen
     ]
+    ++ lib.optionals (hasDtraceSupport && withDtrace) [ systemtap ]
+    ++ lib.optionals parseToYAML [ yq ]
```
The dependency was introduced in https://gerrit.lix.systems/c/lix/+/2190 for some parser tests so that the output is more readable.

alois31 commented on pkgs/tools/package-management/lix/common-nix-eval-jobs.nix:
```diff
@@ -24,11 +25,16 @@ stdenv.mkDerivation {
+      boost
+    ]
+    ++ lib.optionals (lib.versionAtLeast version "2.93") [
+      capnproto
```
The library is already used by Lix itself. This only adds the dependency to nix-eval-jobs which started depending on canproto headers due to the async changes. If a special capnproto build is needed that could be handled similar to boehmgc, aws-sdk-cpp and editline so that it is picked up by both.

alois31 replied:
Correct, I think it might even have been me who suggested making it propagated. I decided to add it to nix-eval-jobs for now at least since that's also done in the vendored packaging.

alois31 commented on pkgs/tools/package-management/lix/common-lix.nix:
```diff
@@ -161,6 +169,8 @@ stdenv.mkDerivation (finalAttrs: {
       mdbook-linkcheck
       doxygen
     ]
+    ++ lib.optionals (hasDtraceSupport && withDtrace) [ systemtap ]
```
Thank you for making me aware that this exists now.

lf- reviewed (approved):
looks good to me, and derisks 2.93 packaging issues. wins all around.

alois31 committed:
lixPackageSets.*.lix: test the correct Lix in passthru.tests

Previously always the top-level lix package would be used, which is obviously
not what we want.

getchoo commented on pkgs/tools/package-management/lix/default.nix:
```diff
@@ -188,7 +202,51 @@ lib.makeExtensible (self: {
   };
 
-  latest = self.lix_2_92;
+  lix_2_93 = self.makeLixScope {
```
Both aarch64/x86_64-darwin are failing here and in the git package, but only the former is a regression from 2.92

It seems this is a repeat of https://github.com/NixOS/nixpkgs/pull/353576, as applying this diff fixed it:

```diff
diff --git a/pkgs/tools/package-management/lix/common-lix.nix b/pkgs/tools/package-management/lix/common-lix.nix
index 13f6d88828e6..635f52082140 100644
--- a/pkgs/tools/package-management/lix/common-lix.nix
+++ b/pkgs/tools/package-management/lix/common-lix.nix
@@ -254,7 +254,8 @@ stdenv.mkDerivation (finalAttrs: {
       # Enable LTO, since it improves eval performance a fair amount
       # LTO is disabled on:
       # - static builds (strange linkage errors)
-      (lib.mesonBool "b_lto" (!stdenv.hostPlatform.isStatic && (isLLVMOnly || stdenv.cc.isGNU)))
+      # - darwin builds (install test failures)
+      (lib.mesonBool "b_lto" (!stdenv.hostPlatform.isStatic && !stdenv.hostPlatform.isDarwin && (isLLVMOnly || stdenv.cc.isGNU)))
       (lib.mesonEnable "gc" true)
       (lib.mesonBool "enable-tests" true)
       (lib.mesonBool "enable-docs" enableDocumentation)
```
It might be best to keep this, as I'm unsure if it's UB from the test's code (from a quick look this seems unlikely) or the LTO-ized Lix binary. The only side effect it would have is retroactively disabling LTO on Lix 2.92 for Darwin (as that's the only place it's enabled currently) where tests do pass; though this might still be an issue for 2.92 and it just isn't being caught

getchoo commented on pkgs/tools/package-management/lix/default.nix:
```diff
@@ -188,7 +202,51 @@ lib.makeExtensible (self: {
+    };
+  };
+
+  latest = self.lix_2_93;
```
Should we set this as stable too? AFAIK there are no upstream or third party blockers like 2.92, and @RaitoBezarius signaled towards skipping over that release anyways [here](https://github.com/NixOS/nixpkgs/pull/405718#issuecomment-2869961790)

RaitoBezarius replied:
I do think we have one major regression, which is the SSH wrapper stuff, atm, that I will be trying to fix in https://gerrit.lix.systems/c/lix/+/3159 and backport it. cc @lf-

getchoo reviewed (commented):
Would be awesome to get this merged for 25.05. Maybe add a release note for it too?

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 398141`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 30 packages built:</summary>
  <ul>
    <li>lixPackageSets.git.aws-sdk-cpp (lixPackageSets.lix_2_93.aws-sdk-cpp)</li>
    <li>lixPackageSets.git.aws-sdk-cpp.dev (lixPackageSets.lix_2_93.aws-sdk-cpp.dev)</li>
    <li>lixPackageSets.git.boehmgc (lixPackageSets.lix_2_93.boehmgc)</li>
    <li>lixPackageSets.git.boehmgc.debug (lixPackageSets.lix_2_93.boehmgc.debug)</li>
    <li>lixPackageSets.git.boehmgc.dev (lixPackageSets.lix_2_93.boehmgc.dev)</li>
    <li>lixPackageSets.git.boehmgc.doc (lixPackageSets.lix_2_93.boehmgc.doc)</li>
    <li>lixPackageSets.git.editline (lixPackageSets.lix_2_93.editline)</li>
    <li>lixPackageSets.git.editline.dev (lixPackageSets.lix_2_93.editline.dev)</li>
    <li>lixPackageSets.git.editline.doc (lixPackageSets.lix_2_93.editline.doc)</li>
    <li>lixPackageSets.git.editline.man (lixPackageSets.lix_2_93.editline.man)</li>
    <li>lixPackageSets.git.lix</li>
    <li>lixPackageSets.git.lix.debug</li>
    <li>lixPackageSets.git.lix.dev</li>
    <li>lixPackageSets.git.lix.devdoc</li>
    <li>lixPackageSets.git.lix.doc</li>
    <li>lixPackageSets.git.lix.man</li>
    <li>lixPackageSets.git.nix-direnv</li>
    <li>lixPackageSets.git.nix-eval-jobs</li>
    <li>lixPackageSets.git.nix-fast-build</li>
    <li>lixPackageSets.git.nix-fast-build.dist</li>
    <li>lixPackageSets.latest.lix (lixPackageSets.lix_2_93.lix)</li>
    <li>lixPackageSets.latest.lix.debug (lixPackageSets.lix_2_93.lix.debug)</li>
    <li>lixPackageSets.latest.lix.dev (lixPackageSets.lix_2_93.lix.dev)</li>
    <li>lixPackageSets.latest.lix.devdoc (lixPackageSets.lix_2_93.lix.devdoc)</li>
    <li>lixPackageSets.latest.lix.doc (lixPackageSets.lix_2_93.lix.doc)</li>
    <li>lixPackageSets.latest.lix.man (lixPackageSets.lix_2_93.lix.man)</li>
    <li>lixPackageSets.latest.nix-direnv (lixPackageSets.lix_2_93.nix-direnv)</li>
    <li>lixPackageSets.latest.nix-eval-jobs (lixPackageSets.lix_2_93.nix-eval-jobs)</li>
    <li>lixPackageSets.latest.nix-fast-build (lixPackageSets.lix_2_93.nix-fast-build)</li>
    <li>lixPackageSets.latest.nix-fast-build.dist (lixPackageSets.lix_2_93.nix-fast-build.dist)</li>
  </ul>
</details>

RaitoBezarius commented:
> Can we go ahead with bumping lixVersions.latest to this while keeping lixVersions.stable at 2.91? Then we can sort out the remaining issues asynchronously.

In agreement with this.
